neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+$$
+\prod e(p_i, q_i) = 1
+$$
+
+We don’t calculate the pairing function itself: the result will be in the huge field, and in all known applications only such a check is necessary.
+
+***Input:*** the sequence of pairs $(p_i, q_i)$, where $p_i \in G_1 \subset E(F_{p})$ and $q_i \in G_2 \subset E'(F_{p^2})$. Each point is encoded in decompressed form. Expected `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the concatenation of uncompressed point from $G_1 \subset E(F_p)$ — `96 bytes` and point from $G_2 \subset E'(F_{p^2})$ — `192 bytes`. More details are in the Curve Points Encoding section.
+
+***Output:*** returns `bool` — the result of the pairing check. The `true` value means that the pairing result is equal to multiplicative identity and `false` otherwise. For empty input it returns `true`.
+
+***Gas Estimation:***
+
+The algorithm has a linear complexity of the number of elements. The gas can be calculated by the next formula
+
+```rust
+let k = (input_bytes + item_size - 1)/item_size
+let gas_consumed = A + B * k
+```
+
+A and B are constants calculated empirically.
+
+Here you can find benchmark test vectors for EIP-2537[^46].
+
+***Test cases:***
+
+- The correct input with different lengths with results true
+- The correct input with different lengths with results false
+- The first point on the curve but not in G1
+- The second point on the curve but not in G2
+- The input with incorrect length
+- The points not on the curve
+- Some points = 0
+- The field elements are encoded incorrectly
+- Empty input
+
+***Tests References:*** The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].
+
+***Error cases:***
+
+- The input length is not divided by 288
+- Too much memory is used
+- Field elements encoded incorrectly (see Curve points encoded section)
+- Any points not on the curve
+- Any points not in $G_1/G_2$
+
+***Annotation:***
+
+```rust
+/// Computes pairing check on BLS12-381 curve.
+/// \prod_i e(g_{1 i}, g_{2 i}) should be equal one, e(g1, g2) is pairing
+///
+/// # Arguments
+///
+/// * `value` - sequence of (g1:G1, g2:G2), where
+///   G2 is subgroup point (x:Fp2, y:Fp2) on BLS12-381 twist,
+///   BLS12-381 twist is Y^2 = X^3 + 4(i + 1) curve over Fp2
+///   Fp2 is complex field element (re: Fp, im: Fp)
+///   G1 is point (x:Fp, y:Fp) on BLS12-381,
+///   BLS12-381 is Y^2 = X^3 + 4 curve over Fp
+///
+///   `value` is encoded a as packed, big-endian
+///   `[(([u8; 48], [u8; 48]), ([u8; 96], [u8; 96]))]` slice.
+///
+/// # Errors
+///
+/// If `value_len + value_ptr` points outside the memory than
+/// the function returns `MemoryAccessViolation`.
+///
+/// If point coordinates are not on curve, point is not in the subgroup
+// or data are wrong serialized, for example,
+/// `value.len()%288!=0`, the function returns `BLS12381InvalidInput`.
```

Comment:
Seems like this is by analogy with current bn128 host functions. It makes me sad (we should have just multiplied by 288 at the beginning of the host function), but I guess consistency is more important than me being happy in this case, so it probably shouldn’t change :’(

Reply:
This doesn't work if the value comes from a register. Maybe we should have never supported using registers to pass input to host functions, but now AFAIK all existing host functions support this.